File: week2/esadmsk_w2/longest-substring-without-repeating-characters.java
 
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLength = 0;
        int left = 0;
        int[] charIndex = new int[256]; 
        
        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);
            left = Math.max(left, charIndex[currentChar]);
            maxLength = Math.max(maxLength, right - left + 1);
            charIndex[currentChar] = right + 1;
        }
        return maxLength;
    }
}

File: week2/MarshalLee/LongestSubstringWithoutRepeatingCharacters.cs
using System;

public class LongestSubstringWithoutRepeatingCharacters
{
    public int LengthOfLongestSubstring(string s)
    {
        HashSet<char> list = new();
        int left = 0, right = 0, maxLength = 0;

        while (right < s.Length)
        {
            if (!list.Contains(s[right]))
            {
                list.Add(s[right]);
                right++;
                maxLength = Math.Max(maxLength, list.Count);
            }
            else
            {
                list.Remove(s[left]);
                left++;
            }
        }
        return maxLength;
    }
}
